secret_store/src/listener/service_contract_listener.rs

```diff
@@ -340,25 +451,86 @@ impl ChainNotify for ServiceContractListener {
 
 impl ClusterSessionsListener<GenerationSession> for ServiceContractListener {
 	fn on_session_removed(&self, session: Arc<GenerationSession>) {
-		// only publish when the session is started by another node
-		// when it is started by this node, it is published from process_service_task
-		if !is_processed_by_this_key_server(&*self.data.key_server_set, &*self.data.self_key_pair, &session.id()) {
-			// by this time sesion must already be completed - either successfully, or not
-			assert!(session.is_finished());
-
-			// ignore result - the only thing that we can do is to log the error
-			match session.wait(Some(Default::default()))
-				.map_err(|e| format!("{}", e))
-				.and_then(|server_key| Self::publish_server_key(&self.data, &session.id(), &server_key)) {
-				Ok(_) => trace!(target: "secretstore", "{}: completed foreign GenerateServerKey({}) request",
-					self.data.self_key_pair.public(), session.id()),
-				Err(error) => warn!(target: "secretstore", "{}: failed to process GenerateServerKey({}) request with: {}",
-					self.data.self_key_pair.public(), session.id(), error),
+		// by this time sesion must already be completed - either successfully, or not
+		assert!(session.is_finished());
+
+		// ignore result - the only thing that we can do is to log the error
+		let server_key_id = session.id();
+		if let Some(origin) = session.origin() {
+			if let Some(generation_result) = session.wait(Some(Default::default())) {
+				let generation_result = generation_result.map(Some).map_err(Into::into);
+				let _ = Self::process_server_key_generation_result(&self.data, origin, &server_key_id, generation_result);
 			}
 		}
 	}
 }
 
+impl ClusterSessionsListener<DecryptionSession> for ServiceContractListener {
+	fn on_session_removed(&self, session: Arc<DecryptionSession>) {
+		// by this time sesion must already be completed - either successfully, or not
+		assert!(session.is_finished());
+
+		// ignore result - the only thing that we can do is to log the error
+		let session_id = session.id();
+		let server_key_id = session_id.id;
+		if let Some(requester) = session.requester().and_then(|r| r.address(&server_key_id).ok()) {
```

Comment:
Couldn't you concatenate all the 3 lines below into 1 `if let` statement? Since they don't depend on the previous value. So like `if let (Some(x), Some(y)) = (session.requester()[..], session.origin())` ?

Reply:
Merged two lines into 1. Not sure this improves readability, though :) 3rd line has side effects => do not want to depend on evaluation order => left separate `if let`